### PR TITLE
feat: search box navigation thru results with up/down keys doesn't work (#3028)

### PR DESCRIPTION
**Pull request by Zeyber**
Closes: #3028

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 1]
```diff
@@ -107,17 +109,29 @@ export class SearchBoxComponent {
...
-    }
-    this.ignoreCloseEvent = false;
+      this.ignoreCloseEvent = false;
+    }, 0);
+  }
+
+  // Check if focus is on searchbox or result list elements
+  private isSearchboxFocused(): boolean {
+    return this.getResultElements().indexOf(this.getFocusedElement()) === -1 &&
+      this.winRef.document.querySelector('input[aria-label="search"]') !==
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: this.winRef.document.querySelector('input[aria-label="search"]') === this.getFocusedElement()

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 2]
```diff
@@ -131,6 +145,57 @@ export class SearchBoxComponent {
     }
   }
 
+  // Return result list as HTMLElement array
+  private getResultElements(): HTMLElement[] {
+    const suggestions = this.winRef.document.querySelectorAll(
+      'div.suggestions > a'
+    );
+    const products = this.winRef.document.querySelectorAll('div.products > a');
+    const results = [];
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: const results: HTMLElement[] = [];

Comment (Pucek9) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: or return [...suggestions, ...products]

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 3]
```diff
@@ -107,17 +109,29 @@ export class SearchBoxComponent {
...
-    }
-    this.ignoreCloseEvent = false;
+      this.ignoreCloseEvent = false;
+    }, 0);
+  }
+
+  // Check if focus is on searchbox or result list elements
+  private isSearchboxFocused(): boolean {
+    return (
+      this.getResultElements().indexOf(this.getFocusedElement()) !== -1 ||
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: !this.getResultElements().includes(this.getFocusedElement())

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 4]
```diff
@@ -131,6 +145,55 @@ export class SearchBoxComponent {
...
+  }
+
+  // Return focused element as HTMLElement
+  private getFocusedElement(): HTMLElement {
+    return <HTMLElement>this.winRef.document.activeElement;
+  }
+
+  // Focus on previous item in results list
+  focusPreviousChild(event) {
+    event.preventDefault(); // Negate normal keyscroll
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]:
```
event.preventDefault(); // Negate normal keyscroll
const results = this.getResultElements();
const focusedIndex = results.indexOf(this.getFocusedElement());
```that part is duplicated so should be separate to function

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 5]
```diff
@@ -131,6 +145,55 @@ export class SearchBoxComponent {
     }
   }
 
+  // Return result list as HTMLElement array
+  private getResultElements(): HTMLElement[] {
+    const suggestions: HTMLElement[] = Array.from(
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: if we don't focus on suggestions and products separately, we can use also 
```
return Array.from(this.winRef.document.querySelectorAll('.products > a, .suggestions > a'));
```

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.html [hunk 6]
```diff
@@ -7,7 +7,11 @@
     (input)="search(searchInput.value)"
     (blur)="close($event)"
     (keydown.escape)="close($event)"
-    (keydown.enter)="launchSearchResult($event, searchInput.value)"
+    (keydown.enter)="
+      launchSearchResult($event, searchInput.value); close($event, true)
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.html]: close is called twice here.
leave `launchSearchResult($event, searchInput.value);`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts (hunk 4 above)
Comment (Pucek9) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: After discuss with Kacper, every approach causes violates the rules of cc, so let leave as is

**Comment by Zeyber** [comment]
QA Steps:
1. Type 'dsa' into searchbox
2. When products or suggestions are loaded, press up key, check highlights last product in list
3. Press up key again, check highlights second last product in list
4. Press down key, check highlights last product in list
5. Press down key again, check highlights first product in list
6. Press enter, check navigates to highlighted option and closes searchbox

**Pull request merged**